File: code_teller/q983/tyama_codeiq983.ex
#!/usr/bin/env elixir
defmodule CodeIQ983 do
	def calc2(d,num,bse) do
		if d==0 do
			rem(num,bse)
		else
			calc2(d-1,div(num,bse),bse)
		end
	end
	def calc(digits,expbase,bse,x,n) do
		if x>n do
			calc2(digits-1-rem(n,digits),expbase+div(n,digits),bse)
		else
			calc(digits+1,expbase*bse,bse,(digits+1)*(expbase*bse)*(bse-1),n-x)
		end
	end
	def main() do
		s=IO.gets('')
		if s != :eof do
			n=elem Integer.parse(s),0
			if n != 0 do
				IO.puts calc(1,1,10,9,n+3)
				main
			end
		end
	end
end
CodeIQ983.main

File: code_teller/q983/tyama_codeiq983.hs
#!/usr/bin/env runghc
import Control.Applicative
import System.IO (isEOF)

calc2 0 num bse =
	mod num bse
calc2 d num bse =
	calc2 (d-1) (div num bse) bse

calc digits expbase bse x n =
	if x>n then
		calc2 (digits-1-(mod n digits)) (expbase+(div n digits)) bse
	else
		calc (digits+1) (expbase*bse) bse ((digits+1)*(expbase*bse)*(bse-1)) (n-x)

main = do
	eof <- isEOF
	if not eof then do
		n<-readLn
		print $ calc 1 1 10 9 (n+3)
		main
	else return ()